Sprint 2

grizzlygarillaz opened the pull request.

grizzlygarillaz committed:
add component helpers

grizzlygarillaz committed:
push sprint_2

grizzlygarillaz committed:
add components

grizzlygarillaz committed:
fix form console output

grizzlygarillaz committed:
fix nanoid incompatible

grizzlygarillaz committed:
fix node version

grizzlygarillaz committed:
fix node version

grizzlygarillaz committed:
fix node version

grizzlygarillaz committed:
fix page routing

grizzlygarillaz committed:
fix page routing

grizzlygarillaz committed:
fix icon output

grizzlygarillaz committed:
fix svg packages

grizzlygarillaz committed:
fix svg packages

grizzlygarillaz committed:
move ts to devDependencies

grizzlygarillaz committed:
fix json's

grizzlygarillaz committed:
add netlify link

eawtf commented on .eslintignore:
```diff
@@ -0,0 +1,3 @@
+/build
+**/*.min.js
```
Мы уже игнорируем директорию `/build`, а где помимо неё у нас есть `.min.js` файлы?
Быть может это правило излишне и уже покрывается правилом `/build`?

eawtf commented on .nvmrc:
```diff
@@ -1 +1 @@
-v12.22.9
\ No newline at end of file
+v12.22.12
```
Почему решено использовать 12-ю версию?
После неё была 14 LTS (long term support) и сейчас уже есть 18, а 16-я в LTS:
<img width="265" alt="image" src="https://user-images.githubusercontent.com/16733354/189531608-96f142ba-cf86-4e52-af7a-f0040933156d.png">


Возможно, стоит использовать 16-ю версию?
(новые версии часто несут более свежий и корректный функционал, исправленные ошибки)

Или есть определенные проблемы с совместимостью?

grizzlygarillaz replied:
Стояла свежая версия, но т.к. гит проверял по 12-ой выдавал ошибку на некоторые модули и, чтобы точно пройти тесты поставил себе идентичную с тестами версию

eawtf commented on package.json:
```diff
@@ -5,19 +5,33 @@
+    "postcss-sass": "0.5.0"
+  },
+  "devDependencies": {
+    "@parcel/transformer-inline-string": "^2.7.0",
```
В `package.json` опять попали нефиксированные версии библиотек.

Символ '^' указывает что пакетный менеджер может взять "совместимую" версию библиотеки, у которой, например, минорная версия (1.2.3) или версия патча (1.2.3) выше текущей указанной версии.

И может быть случай, например, что вы с командой удалите из репозитория файл package-lock.json и попытаетесь повторно установить зависимости - может случиться так, что установятся другие версии зависимостей, выше версии, используемой ранее. А это может привести к ошибкам, так как новая функциональность может отличаться от той, которая ожидалась.

Так что, хорошая практика и рекомендация - использовать фиксированные версии и обновлять их в рамках соответствующих задачек отдельно, протестировав, что всё работает корректно.

Давай заменим здесь и ниже все версии на фиксированные.

Полезные ссылки:
https://github.com/npm/node-semver#caret-ranges-123-025-004

![image](https://user-images.githubusercontent.com/16733354/189175284-b465be87-634e-417d-bda1-5c56a3d1a7d2.png)

eawtf commented on package.json:
```diff
@@ -5,19 +5,33 @@
   "license": "ISC",
   "dependencies": {
     "@anikin/parcel-transformer-handlebars-precompile": "1.0.1",
     "@parcel/transformer-sass": "2.7.0",
```
Точно ли данная библиотека должна находиться в секции `dependencies`?
Или же мы можем перенести её в `devDependencies`?

<img width="762" alt="image" src="https://user-images.githubusercontent.com/16733354/189175943-cd281440-9eca-4621-923c-e6d2537fb1b7.png">

Полезные ссылки:
https://habr.com/ru/company/domclick/blog/509440/

eawtf commented on src/components/Button/type.ts:
```diff
@@ -0,0 +1,8 @@
+  label:string,
+  onClick: () => void,
+  events: Record<string, () => void>
+  type:string,
```
Хороший тон разработки - соблюдение "код-стайла" и единообразия написания.
В каких-то местах здесь у нас есть пробел между пропом и его типом, а где-то нет.

Давай сделаем единообразно, добавив пробел везде (у `type` и `label` тоже).

eawtf commented on src/components/Input/input.hbs:
```diff
@@ -0,0 +1,6 @@
+<input type="{{ type }}"
```
Может быть добавить немного читаемости, немного отформатировав отступы как-нибудь вроде:

```
<input type="{{ type }}"
   class="{{className}}"
   name="{{ name }}"
   value="{{ value }}"
   title="{{ title }}"
   
   {{#if required}}
     required
   {{/if}}
 />
```
    
  (на своё усмотрение)

eawtf commented on src/components/InputBox/index.ts:
```diff
@@ -0,0 +1,56 @@
+          this.toggleError();
+        },
+        focusin: () => {
+          this.toggleError();
```
А точно ли мы хотим показывать ошибку на фокус с точки зрения UX?
Пользователя кликнул на инпут - хочет что-то ввести, зачем ему показывать ошибку, он сейчас будет менять, потом делает blur (focusout) и вот тут уже можно проверить - а корректно ли он ввёл.

Но это тоже, вопрос к дизайну приложения уже, не критично :)

eawtf commented on src/util/HTTPTransport.ts:
```diff
@@ -0,0 +1,102 @@
+
+    if (data) {
+      Object.keys(data).forEach((key: keyof HttpBody) => {
+        const value = Array.isArray(data[key]) ? (data[key] as []).join(',') : data[key];
```
А что если у нас `data[key]` является непримитивов, а, например, объектом?
Будет ли данный функционал работать?

Может обернуть `data[key]` в `JSON.stringify(...)`?

eawtf commented on src/util/Validator.ts:
```diff
@@ -0,0 +1,29 @@
+  return true;
+}
+
+export default Validator;
```
Хорошая реализация 🔥

eawtf commented:
Отличная работа, очень понравилось - чувствуется, что есть понимание и желание писать правильный код.

Давай пробежимся по комментариям и поправим где нужно небольшие мелочи (например, те же версии библиотек) и думаю, что можно мерджить 👍 

Спасибо 🔥

grizzlygarillaz committed:
review fixes
(.eslintignore) remove unnecessary rule
(package.json) replace versions with fixed
(package.json) move dev packages to devDep
(codestyle) fix few mistakes at .ts interfaces
(codestyle) make Input component more readable
(InputBox) remove focusin validation
(HTTPTransport) fix when data array can contain a non primitive value

eawtf reviewed (commented):
Отлично, теперь выглядит ещё лучше 👍 
Здорово, аппрув, можем мерджить ✅
